Add Aero Offboard velocity example

padmashree9110 opened the pull request:
Demonstrates Aero Offboard velocity control using both NED and Body co-ordinates.

padmashree9110 committed:
Add Aero Offboard velocity example

Demonstrates Aero Offboard velocity control using both NED and Body co-ordinates.

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/include/aero_drone_action.h:
```diff
@@ -0,0 +1,45 @@
+#include <sensor_msgs/NavSatFix.h>
+#include <mavros_msgs/HomePosition.h>
+#include <std_msgs/Float64.h>
+#include <mavros_msgs/Altitude.h>
```
Add a line break.

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/include/aero_drone_action.h:
```diff
@@ -0,0 +1,45 @@
+  bool land();
+  bool setOffboardMode();
+
+  // velocity
```
Add comment as: **Offboard velocity control**

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/include/aero_drone_action.h:
```diff
@@ -0,0 +1,45 @@
+  bool setOffboardMode();
+
+  // velocity
+  void setOffboardVelocityNed(float vx, float vy, float vz, float yaw);
```
Capitalize the entire word, **NED** in all occurrences.

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/include/aero_drone_action.h:
```diff
@@ -0,0 +1,45 @@
+  ros::Publisher set_vel_pub_;
+  ros::ServiceClient takeoff_client_;
+  ros::ServiceClient set_mode_client_;
+  ros::ServiceClient land_client_;
```
Group data member & methods logically for readability, as below:
```
void getHomeGeoPoint();
void getAltitude();
void setHomeGeoPointCB(const mavros_msgs::HomePositionConstPtr& home);
void getAltitudeCB(const mavros_msgs::AltitudeConstPtr& altitude);
float toRadFromDeg(float deg);

ros::ServiceClient takeoff_client_;
ros::ServiceClient set_mode_client_;
ros::ServiceClient land_client_;
```

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/package.xml:
```diff
@@ -0,0 +1,19 @@
+
+<maintainer email="padmashree.mandri@intel.com">Padmashree Mandri</maintainer>
+
+<license>TODO</license>
```
Add license as **BSD 3-Clause**

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/src/aero_drone_action.cpp:
```diff
@@ -0,0 +1,188 @@
+                  mavros_msgs::PositionTarget::IGNORE_PZ | mavros_msgs::PositionTarget::IGNORE_AFX |
+                  mavros_msgs::PositionTarget::IGNORE_AFY | mavros_msgs::PositionTarget::IGNORE_AFZ |
+                  mavros_msgs::PositionTarget::FORCE | mavros_msgs::PositionTarget::IGNORE_YAW_RATE;
+  pos.position.x = 0.0f;
```
Remove explicit initialization of `0.f` and compensate it by using [value initialization](http://en.cppreference.com/w/cpp/language/value_initialization) method as below:
```
mavros_msgs::PositionTarget pos{};
```

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/src/aero_drone_action.cpp:
```diff
@@ -0,0 +1,188 @@
+  set_vel_pub_.publish(pos);
+}
+
+void AeroDrone::setOffboardVelocityNed(float vx, float vy, float vz, float yaw, int count)
```
Make `count` as `unsigned int` or equivalent.

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/src/aero_offboard_velocity_node.cpp:
```diff
@@ -0,0 +1,205 @@
+/**
+ * @file mavros_offboard_velocity_node.cpp
+ * @brief MAVROS Offboard Control example node, written with mavros version 0.14.2, px4 flight
```
Update MAVROS version to whatever you've tested on.

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/src/aero_offboard_velocity_node.cpp:
```diff
@@ -0,0 +1,205 @@
+/**
+ * @file mavros_offboard_velocity_node.cpp
+ * @brief MAVROS Offboard Control example node, written with mavros version 0.14.2, px4 flight
+ * stack and tested in Gazebo SITL & jMAVSIM.
```
You may remove **jMAVSIM** if you couldn't verify on it. Good to add version info of PX4 & Gazebo.

padmashree9110 replied:
Thanks for suggestion shakthi. But it works fine with jmavsim.

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/src/aero_offboard_velocity_node.cpp:
```diff
@@ -0,0 +1,205 @@
+    ROS_INFO("OFFBOARD enabled");
+  else
+  {
+    ROS_INFO("unable to switch to offboard");
```
Use `ROS_ERROR()` for error logs, wherever applicable & capitalize start letter of them.

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/src/aero_offboard_velocity_node.cpp:
```diff
@@ -0,0 +1,205 @@
+  ROS_INFO("NED: Move south");
+  aero.setOffboardVelocityNed(0.0f, -2.0f, 0.0f, 180.0f, 200);
+
+  //Wait for a bit
```
Add a SPACE after `//`

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/src/aero_offboard_velocity_node.cpp:
```diff
@@ -0,0 +1,205 @@
+ */
+
+#include <cstdlib>
+#include "aero_drone_action.h"
```
Add a line break.

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/src/aero_offboard_velocity_node.cpp:
```diff
@@ -0,0 +1,205 @@
+  //Wait for a bit
+  aero.setOffboardVelocityNed(0.0f, 0.0f, 0.0f, 0.0f, 20);
+
+  ROS_INFO("NED: Move south");
```
Add it as _"Move towards South"_ wherever applicable.

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/src/aero_offboard_velocity_node.cpp:
```diff
@@ -0,0 +1,205 @@
+  ros::init(argc, argv, "offb_node");
+  AeroDrone aero;
+  int ret = EXIT_SUCCESS;
+  bool ret1;
```
Use a meaningful name: `is_success` for eg.

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/src/aero_offboard_velocity_node.cpp:
```diff
@@ -0,0 +1,205 @@
+  sleep(5);  // Let Aero reach takeoff altitude
+
+  // using local NED co-ordinates
+  ret1 = offboardCtrlNed(aero);
```
`offboardCtrlNed` => `offboardVelocityCtrlNED`

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/src/aero_offboard_velocity_node.cpp:
```diff
@@ -0,0 +1,205 @@
+  }
+
+  // using body co-ordinates
+  ret1 = offboardCtrlBody(aero);
```
`offboardCtrlBody` => `offboardVelocityCtrlBody`

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/src/aero_offboard_velocity_node.cpp:
```diff
@@ -0,0 +1,205 @@
+
+  // using local NED co-ordinates
+  ret1 = offboardCtrlNed(aero);
+  if (ret1 == true)
```
For boolean variables, you can simply say `if (is_success) {...}`, which is more readable than explicit comparison.

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/src/aero_offboard_velocity_node.cpp:
```diff
@@ -0,0 +1,205 @@
+  ret1 = offboardCtrlNed(aero);
+  if (ret1 == true)
+  {
+    ROS_INFO("Done with Ned velocity\n");
```
Doesn't sound that we're done with Offbaord Velocity control using NED coordinates. It can be little more descriptive.

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/src/aero_offboard_velocity_node.cpp:
```diff
@@ -0,0 +1,205 @@
+  ret1 = offboardCtrlBody(aero);
+  if (ret1 == true)
+  {
+    ROS_INFO("Done with body velocity");
```
Same as [above](https://github.com/intel-aero/sample-apps/pull/12/commits/eb2463d22a963ffff0eb57f349ba4a29e6f50828#r161673111).

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/src/aero_offboard_velocity_node.cpp:
```diff
@@ -0,0 +1,205 @@
+  sleep(5);
+end:
+  ros::shutdown();
+  return 0;
```
Return as `return ret;`, I believe that was the purpose of capturing return result to `ret`. Isn't it ?

shakthi-prashanth-m reviewed (changes requested):
Thanks @padmashree9110  for submitting Offboard velocity control example. Please consider the changes I have requested.

padmashree9110 committed:
Improve Aero Offboard Velocity example

1>Added stepUp and stepDown functions
2>Updated Mavros Version
3>Other minor changes

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/src/aero_offboard_velocity_node.cpp:
```diff
@@ -1,125 +1,196 @@
+
+void stepUp(AeroDrone &aero);
+void stepDown(AeroDrone &aero);
+bool offboardVelocityCtrlNED(AeroDrone &aero);
```
Can you please add `offboardVelocityCtrlNED()` and `offboardVelocityCtrlBody()` first ? As `stepUp()` and `stepDown()` are utility methods for the former ones.
```
bool offboardVelocityCtrlNED(AeroDrone &aero);
bool offboardVelocityCtrlBody(AeroDrone &aero);

// Utility methods
void stepUp(AeroDrone &aero);
void stepDown(AeroDrone &aero);
```

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/src/aero_offboard_velocity_node.cpp:
```diff
@@ -1,125 +1,196 @@
+  return ret;
+}
 
+void stepUp(AeroDrone &aero)
```
Maintain the same order as said [here](https://github.com/intel-aero/sample-apps/pull/12/files/eb2463d22a963ffff0eb57f349ba4a29e6f50828..716933ce7cdb5e925114caeb471d35bbbc7703cf#r162000521).

shakthi-prashanth-m commented on ros/src/aero_offboard_velocity/src/aero_offboard_velocity_node.cpp:
```diff
@@ -1,125 +1,196 @@
+  if (is_success)
+  {
+    ROS_INFO("Completed Offboard Velocity Control with NED co-ordinates\n");
+  }
```
Perform error handling for Offboard velocity control (both NED & Body).

shakthi-prashanth-m reviewed (commented):
Its very much better now :+1:  Please resolve minor changes requested.

padmashree9110 committed:
Improve Aero Offboard Velocity example

1>Added px4 FW vesrion and gazebo version
2>Other minor changes

padmashree9110 committed:
Improve Aero Offboard Velocity example

1>Replaced c-style type cast by that of c++
2>Other minor changes

shakthi-prashanth-m commented:
@anitha-suresh I am done with the review . You can review and consider merging this commit.

anitha-suresh-zz commented on ros/src/aero_offboard_velocity/src/aero_drone_offboard.cpp:
```diff
@@ -0,0 +1,184 @@
+  return static_cast<float>(deg / 180.0f * M_PI);
+}
+
+void AeroDrone::setOffboardVelocityBody(float vx, float vy, float vz, float yaw_rate)
```
I think we don't need to have two different functions just for 'count' parameter.

shakthi-prashanth-m replied:
OK. We'd two different methods to ease developer usage (every time passing 1 is mandatory for setting single offboard command).
However, we can use default arguments.
@padmashree9110 you can use as below:
```
void
AeroDrone::setOffboardVelocityBody(float vx, float vy, float vz,
                                   float yaw_rate,
                                   size_t count = 1);
// count 1 is passed by default; can be overridden
```

anitha-suresh-zz commented on ros/src/aero_offboard_velocity/src/aero_drone_offboard.cpp:
```diff
@@ -0,0 +1,184 @@
+
+bool AeroDrone::setOffboardMode()
+{
+  mavros_msgs::SetMode offb_set_mode;
```
rename offb to offboard

padmashree9110 committed:
Improve Aero offboard Velocity

1>Modified functions
2>Other minor changes
